### PR TITLE
Local assembly

### PR DESCRIPTION
**Pull request by Rombur**
The main commit is the fifth one where the new functions are added. I have checked that when one eigenvalue has a multiplicity greater than one the eigenvectors produced by arpack are orthogonal.

I have also added a new template parameter `NumberType`. This is not strictly necessary because we can get away with template deduction in the functions where `NumberType` appears. However, it makes it easier for ETI. So I would like to keep it for now and change this later.

Relates to #8

### CODE REVIEW

File: cmake/SetupMFMG.cmake [hunk 1]
```diff
@@ -7,8 +7,8 @@ SET(MFMG_VERSION
 MESSAGE("-- mfmg version: ${MFMG_VERSION}")
 
 # Set flags for Release and Debug version
-set(CMAKE_CXX_FLAGS_RELEASE "-O3")
-set(CMAKE_CXX_FLAGS_DEBUG "-g")
+SET(CMAKE_CXX_FLAGS_RELEASE "-O3")
+SET(CMAKE_CXX_FLAGS_DEBUG "-g")
```
[the remark this answers is not in the document]
Comment (Rombur) [reply cmake/SetupMFMG.cmake]: Is it what the documentation [here](https://cmake.org/Wiki/CMake_Useful_Variables#Compilers_and_Tools) means?

[the remark this answers is not in the document]
Comment (aprokop) [reply cmake/SetupMFMG.cmake]: Kind of. They don't say exactly what the flags are, just stating:
```
You can use these default compilation flags (or modify them) by setting the 
CMAKE_BUILD_TYPE variable at configuration time on the command line, or from within the 
"ccmake" GUI. Note! The default values for these flags change with different compilers. If 
CMake does not know your compiler, the contents will be empty. 
```

[the remark this answers is not in the document]
Comment (Rombur) [reply cmake/SetupMFMG.cmake]: Yeah it is kind of confusing what they mean exactly :/

[the remark this answers is not in the document]
Comment (aprokop) [reply cmake/SetupMFMG.cmake]: In any case, we should either use the default flags, or set our own `CMAKE_CXX_FLAGS_RELEASE` and not explicitly touch `CMAKE_CXX_FLAGS` in cmake infrastructure, only in configuration scripts.

**Comment by Rombur** [comment]
Done

**Pull request merged**